camd.analysis.GeneraticATFAnalyzer

alfredliu-TRI opened the pull request:
1. Implemented a genetic ATF analyzer which takes exploration df and seed df collected from CAMD campaigns. 
`class GenericATFAnalyzer`
2. Four active learning metrics, anyALM, allALM, deALM, and simALM are implemented. 
`gather_deALM(self, exploration_df, record_list)`
`gather_anyALM(self, exploration_df, record_list, percentile=0.01)`
`gather_allALM(self, exploration_df, record_list, percentile=0.1)`
`gather_simALM(self, record_list)`
3. An example is provided in the example folder under analysis_GeneraticATFAnalyzer, together with example data. 

@LindaHung-TRI @SantoshSuram-TRI @JosephMontoya-TRI

JosephMontoya-TRI reviewed (changes requested):
This looks pretty solid on a first pass.  Two big things before I dig into it though - 

1. New functionality needs to be covered by unittests.  You can write a new one, or repurpose some of the example code into a test, this should resolve the 2.6% drop in coverage.
2. Be careful adding whole directories to git, you'll end up adding stuff like .DS_store files unintentionally.  I recommend in general using `git add -p` to track new changes.  We should be ignoring these using `.gitignore` anyway, so go ahead and add `**/.DS_Store
` to our top-level `.gitignore`.

alfredliu-TRI commented:
> This looks pretty solid on a first pass. Two big things before I dig into it though -
> 
> 1. New functionality needs to be covered by unittests.  You can write a new one, or repurpose some of the example code into a test, this should resolve the 2.6% drop in coverage.
> 2. Be careful adding whole directories to git, you'll end up adding stuff like .DS_store files unintentionally.  I recommend in general using `git add -p` to track new changes.  We should be ignoring these using `.gitignore` anyway, so go ahead and add `**/.DS_Store ` to our top-level `.gitignore`.

@JosephMontoya-TRI 
I added the test function, also modified the .gitignore. Now the pull request should point to the latest commit to my master branch.

JosephMontoya-TRI commented on camd/analysis.py:
```diff
@@ -749,3 +749,193 @@ def update_run_w_structure(folder, hull_distance=0.2, parallel=True):
                     )
+
+
+class GenericATFAnalyzer:
```
If this is to be an analyzer, we should subclass the AnalyzerBase abstract class and implement the `analyze` method, otherwise it won't be compatible with the generic CAMD campaign.  I think the logic in your example in `run_analysis_alm` can be adapted into an analyze method, which should take in experimental results from a campaign (in this case, a meta-agent campaign) and seed data (results from past campaigns) and return a summary and a new seed.

JosephMontoya-TRI replied:
You can look at `campaigns.meta_agent.StabilityCampaignAnalyzer` for an example of an analyzer that analyzes campaign results

alfredliu-TRI replied:
I reformatted the `run_analysis_alm` in example into an `analyze` in `GeneraticATFAnalyzer`

JosephMontoya-TRI commented on camd/analysis.py:
```diff
@@ -749,3 +749,193 @@ def update_run_w_structure(folder, hull_distance=0.2, parallel=True):
+
+class GenericATFAnalyzer:
+    """
+        Generic analyzer provide AL metrics analysis, including
```
De-indent these to match the rest of the CAMD doc style.  Sorry pydocstyle's not catching that, will look into configuring it to do so in the future.

alfredliu-TRI replied:
OK, done

JosephMontoya-TRI commented on camd/tests/test_analysis.py:
```diff
@@ -67,5 +68,45 @@ def test_analyze_vaspqmpy_jobs(self):
+        
+        # test deALM
+        deALM = analyzer.gather_deALM(exploration_df, record_list)
+        assert deALM.shape[0] == num_of_agents
```
Rather than inline assert statements, please use the `unittest.TestCase` instance methods, e.g. `self.assertEqual(value1, value2)` and others.  these help when debugging the tests by showing more information about the test failures.

alfredliu-TRI replied:
Done

MuratAykol-TRI commented:
@alfredliu-TRI fyi, you still have the DS_store files in your branch. Also, the example is very useful, but should remove the raw outputs in examples/analysis_GeneraticATFAnalyzer/camd_campaigns/ from git

JosephMontoya-TRI reviewed (changes requested):
A few comments, the biggest one is that in general the logic is all there I think but we need to organize for CAMD compatibility.  There are also remaining .DS_store files and test artifacts, please remove these.  Basically anything that is generated by the code and is a-priori unnecessary to the running of the tests/code.  This helps us keep the repo as simple as possible to minimize download overhead and ensure maintainability.

JosephMontoya-TRI commented on examples/analysis_GeneraticATFAnalyzer/analysis_GeneraticATFAnalyzer.py:
```diff
@@ -0,0 +1,166 @@
+    return record_list
+
+
+def run_analysis_alm(exploration_space_df, record_list):
```
I think we should refactor this example to invoke a Campaign and run the analysis within the campaign, but let's save that for another PR.

JosephMontoya-TRI reviewed (changes requested):
Looks good!  Almost there, one last thing and we can merge.

JosephMontoya-TRI commented:
LGTM.
